src/transport/macfuse/session.rs

```diff
@@ -0,0 +1,527 @@
+        let fd = self.file.as_raw_fd();
+        loop {
+            match read(fd, &mut self.buf) {
+                Ok(len) => {
+                    // ###############################################
+                    // Note: it's a heavy hack to reuse the same underlying data
+                    // buffer for both Reader and Writer, in order to reduce memory
+                    // consumption. Here we assume Reader won't be used anymore once
+                    // we start to write to the Writer. To get rid of this hack,
+                    // just allocate a dedicated data buffer for Writer.
+                    let buf = unsafe {
+                        std::slice::from_raw_parts_mut(
+                            self.buf.as_mut_ptr(),
+                            self.buf.len(),
+                        )
+                    };
+                    // Reader::new() and Writer::new() should always return success.
+                    let reader =
+                        Reader::new(FuseBuf::new(&mut self.buf[..len])).unwrap();
+                    let writer = Writer::new(fd, buf).unwrap();
+                    return Ok(Some((reader, writer)));
+                }
+                Err(e) => match e {
+                    Errno::ENOENT => {
+                        // ENOENT means the operation was interrupted, it's safe
+                        // to restart
+                        trace!("restart reading");
+                        continue;
+                    }
+                    Errno::EAGAIN | Errno::EINTR => {
+                        continue;
+                    }
+                    Errno::ENODEV => {
+                        info!("fuse filesystem umounted");
+                        return Ok(None);
+                    }
+                    e => {
+                        warn! {"read fuse dev failed on fd {}: {}", fd, e};
+                        return Err(SessionFailure(format!(
+                            "read new request: {:?}",
+                            e
+                        )));
+                    }
+                },
+            }
+        }
+    }
+}
+
+/// Safe wrapper for `sysconf(_SC_PAGESIZE)`.
+#[inline(always)]
+fn pagesize() -> usize {
+    // Trivially safe
+    unsafe { sysconf(_SC_PAGESIZE) as usize }
+}
+
+/// Mount a fuse file system
+fn fuse_kern_mount(mountpoint: &Path, fsname: &str, subtype: &str) -> Result<File> {
+    unsafe {
+        let mut mountpoint_argv = String::from(mountpoint.to_str().unwrap());
+        let mut mountpoint_argv = CString::new(mountpoint_argv).unwrap();
+
+        let mut argv = vec![
+            mountpoint_argv.into_raw(),
+        ];
+        // let mut argv = ManuallyDrop::new(argv);
+        let mut args = macfuse::fuse_args {
+            argc: argv.len() as i32,
+            argv: argv.as_mut_ptr(),
+            allocated: 0,
+        };
+
+        let fd = macfuse::fuse_mount_compat25(
```

Comment:
macfuse::fuse_mount_compat25 depend on c libfuse, maybe we can port this method into rust in the future